api/python/provisioner/commands/get_setup_info.py

```diff
@@ -278,14 +269,92 @@ def _get_storage_type(self):
             res[config.STORAGE_TYPE] = config.StorageType.ENCLOSURE.value
         elif popular_revision.startswith("S"):
             # Indium controller type. For example, S100
-            res[config.STORAGE_TYPE] = config.StorageType.RBOD.value
-        else:
+            res[config.STORAGE_TYPE] = config.StorageType.PODS.value
+        elif popular_revision.startswith("E"):
+            # For example, E002
+            # TODO: Do we always have revision starts with "E" for JBOD?
+            #  What other variants are possible?
             res[config.STORAGE_TYPE] = config.StorageType.JBOD.value
+        else:
+            res[config.STORAGE_TYPE] = config.StorageType.OTHER.value
 
         # TODO: EOS-12418-improvement: How to determine EBOD?
 
         return res
 
+    def _get_storage_type(self):
+        """
+        Get storage type
+
+        Algorithm:
+        1. take storage_type/type pillar value. Filled from config.ini
+        2. If previous pillar value is not defined, use storage_type detection
+            method based on lsscsi command output
+        3. take storage_type/controller/type pillar value and convert it to
+            storage_type
+        :return:
+        """
+        def _update_storage_type_pillar(storage_type: str):
+            """
+            Update storage_type pillar by new value
+
+            :param storage_type: new storage_type value for pillar
+                                storage_enclosure/type
+            :return:
+            """
+            _storage_enclosure_path = KeyPath('storage_enclosure')
+            _storage_enclosure_type = PillarKey(
+                                        _storage_enclosure_path / 'type')
+            pillar_updater = PillarUpdater(config.ALL_MINIONS)
+            pillar_updater.update((_storage_enclosure_type, storage_type))
+
+        # Get storage type from pillar values
+        try:
+            return self._get_storage_type_pillar_based()
+        except BadPillarDataError:
+            logger.debug("Pillar based approach for storage_type detection "
+                         "failed")
+
+        # detect storage type using info about system configuration
+        _res = self._detect_storage_type()
+
+        if _res.get(config.STORAGE_TYPE) != config.StorageType.OTHER.value:
+            _update_storage_type_pillar(_res.get(config.STORAGE_TYPE))
+            return _res
+
+        logger.debug("Storage smart detection function returned storage_type "
+                     f"'{_res[config.STORAGE_TYPE]}'")
+
+        # try to detect storage_type using the original method via controller
+        # type pillar value
+        res = dict()
+        storage_enclosure_path = KeyPath('storage_enclosure')
```

Comment:
Please move that to a separate method. Update fun might be separated as a class nethod as well. So here you would have only upper level logic. More modular and clear.

Reply:
Fixed.